crates/syn2mas/src/synapse_reader.rs

```diff
@@ -0,0 +1,281 @@
+//! # Synapse Database Reader
+//!
+//! This module provides facilities for streaming relevant types of database records from a Synapse database.
+
+use async_stream::stream;
+use chrono::{DateTime, Utc};
+use futures_util::Stream;
+use sea_query::{enum_def, Expr, Iden, PostgresQueryBuilder, Query};
+use sea_query_binder::SqlxBinder;
+use sqlx::{query, query_with, FromRow, PgConnection, Postgres, Row, Type};
+use thiserror::Error;
+use thiserror_ext::ContextInto;
+
+#[derive(Debug, Error, ContextInto)]
+pub enum Error {
+    #[error("database error whilst {context}: {source}")]
+    Database {
+        source: sqlx::Error,
+        context: String,
+    },
+}
+
+#[derive(Clone, Debug, sqlx::Decode)]
+pub struct FullUserId(pub String);
+
+impl Type<Postgres> for FullUserId {
+    fn type_info() -> <sqlx::Postgres as sqlx::Database>::TypeInfo {
+        <String as Type<Postgres>>::type_info()
+    }
+}
```

Comment:
We *generally* go for using the database type during lookup, and then convert (usually through a `TryFrom` impl), but I think in this case it makes sense, as Synapse has quite a few recurring patterns like timestamps as ints, booleans as ints, etc.